prov/util/src/util_mr_cache.c

```diff
@@ -0,0 +1,1333 @@
+			if (OFI_UNLIKELY(grc != FI_SUCCESS))
+				FI_WARN(&core_prov, FI_LOG_MR,
+					"failed to release reference to parent, "
+					"parent=%p refs=%d\n",
+					parent, ofi_atomic_get32(&parent->ref_cnt));
+		}
+
+		ofi_atomic_dec32(&cache->inuse.elements);
+
+		if (!util_entry_is_retired(entry)) {
+			iter = rbtFind(cache->inuse.rb_tree, &entry->key);
+			if (OFI_UNLIKELY(!iter)) {
+				FI_WARN(&core_prov, FI_LOG_MR,
+						"failed to find entry in the inuse cache\n");
+			} else {
+				rc = rbtErase(cache->inuse.rb_tree, iter);
+				if (OFI_UNLIKELY(rc != RBT_STATUS_OK)) {
+					FI_WARN(&core_prov, FI_LOG_MR,
+						 "failed to erase lru entry"
+						 " from stale tree\n");
+				}
+			}
+		}
+
+		/* if we are doing lazy dereg and the entry
+		 * isn't retired, put it in the stale cache
+		 */
+		if (cache->attr.lazy_deregistration && !(util_entry_is_retired(entry))) {
+			FI_DBG(&core_prov, FI_LOG_MR,
+			       "moving key %"PRIu64":%"PRIu64" to stale (entry %p)\n",
+			       entry->key.address, entry->key.length, entry);
+
+			found = rbtFindLeftmost(cache->stale.rb_tree,
+						&entry->key,
+						util_find_overlapping_addr);
+			if (found) {
+				/* one or more stale entries would overlap with this
+				 * new entry. We need to resolve these collisions by dropping
+				 * registrations
+				 */
+				util_resolve_stale_entry_collision(cache, found, entry);
+			} else {
+				/* if not found, ... */
+				grc = util_insert_entry_into_stale(cache, entry);
+			}
+		} else {
+			/* if retired or not using lazy registration */
+			FI_DBG(&core_prov, FI_LOG_MR,
+			       "destroying entry, key=%"PRIu64":%"PRIu64" (entry %p)\n",
+			       entry->key.address, entry->key.length, entry);
+
+			grc = util_mr_cache_entry_destroy(cache, entry);
+		}
+
+		if (OFI_UNLIKELY(grc != FI_SUCCESS))
+			FI_INFO(&core_prov, FI_LOG_MR,
+				"dereg callback returned %d\n", grc);
+	}
+
+
+	return grc;
+}
+
+/**
+ * Checks the sanity of cache attributes
+ *
+ * @param[in]   attr  attributes structure to be checked
+ * @return      FI_SUCCESS if the attributes are valid
+ *              -FI_EINVAL if the attributes are invalid
+ */
+static inline int
+util_check_mr_cache_attr_sanity(struct util_mr_cache_attr *attr)
+{
+	/* 0 < attr->hard_reg_limit < attr->soft_reg_limit */
+	if (attr->hard_reg_limit > 0 &&
+			attr->hard_reg_limit < attr->soft_reg_limit)
+		return -FI_EINVAL;
+
+	/* callbacks must be provided */
+	if (!attr->reg_callback || !attr->dereg_callback)
+		return -FI_EINVAL;
+
+	/* valid otherwise */
+	return FI_SUCCESS;
+}
+
+int ofi_util_mr_cache_init(struct util_mr_cache **cache,
+			   struct util_mr_cache_attr *attr)
+{
+	struct util_mr_cache_attr *cache_attr =
+		&default_mr_cache_attr;
+	struct util_mr_cache *cache_p;
+	int rc;
+
+	/* if the provider asks us to use their attributes, are they sane? */
+	if (attr) {
+		if (util_check_mr_cache_attr_sanity(attr) != FI_SUCCESS)
+			return -FI_EINVAL;
+
+		cache_attr = attr;
+	}
```

Comment:
Why not? Sanity checking of input variables seems like something we should be doing here during the initialization. If the structure is not capable of handling the values, or the values are contradictory due to a bug in the provider code, shouldn't we catch it here rather than letting it behave poorly or crash?

Reply:
I've droped the default attributes. 
I guess that the sanity checks is need to be here.